File: web/src/userMaps/transform/gcpMesh.ts
import { resolveSourceRect } from "../sourceRect";
import type { PixelRect } from "../types";
import { applyAffine, type AffineParams } from "./affine";
import type { LatLngPoint, PixelSize } from "./projection";
import { applyTps, type TpsParams } from "./tps";
import { fromMercator } from "./webMercator";

/**
 * One cell, two triangles — and that is exact, not an approximation.
 *
 * Leaflet's screen space is Web Mercator scaled and translated, so a
 * pixel->Mercator affine composes with it into a map that is still affine all
 * the way to the canvas. There is no curvature for a denser lattice to
 * absorb. (The embedded-GeoTIFF path uses 8x8 because it goes
 * pixel->UTM->WGS84->Mercator, and UTM->Mercator genuinely curves.)
 *
 * It is also the performance answer: the georeferencer re-solves on every
 * pointer move during a GCP drag, and each mesh cell costs two clipped
 * drawImage calls over the full preview. Two triangles per frame instead of
 * 128 is the difference between a live warp and a slideshow.
 */
export const AFFINE_GRID_SIZE = 1;

/**
 * Lattice of geographic positions over the raster, in the same row/col order
 * as `buildLatLngMesh` in `projection.ts` (row = pixel Y, col = pixel X) so
 * `WarpedRasterLayer` can consume either without caring which produced it.
 */
export function buildGcpLatLngMesh(
  params: AffineParams,
  pixelSize: PixelSize,
  gridSize: number = AFFINE_GRID_SIZE,
  sourceRect?: PixelRect,
): LatLngPoint[][] {
  const rect = resolveSourceRect(pixelSize, sourceRect);
  const mesh: LatLngPoint[][] = [];
  for (let row = 0; row <= gridSize; row += 1) {
    const line: LatLngPoint[] = [];
    const y = rect.y + (rect.height * row) / gridSize;
    for (let col = 0; col <= gridSize; col += 1) {
      const x = rect.x + (rect.width * col) / gridSize;
      line.push(fromMercator(applyAffine(params, x, y)));
    }
    mesh.push(line);
  }
  return mesh;
}

/**
 * Coarse tier for a TPS mesh while a control point is actively being dragged.
 *
 * At gridSize 16 the mesh error is 44.3/10.9/15.9 ground m (max, three real
 * Church control sets) and re-solving stays inside a 16 ms frame up to
 * n = 300 control points (measured: solve + evaluate at gridSize 12/16 is
 * the n = 300 ceiling; at gridSize 64 it drops to n = 200). That is what
 * makes 16 the drag-time choice — it is cheap enough to redraw on every
 * pointer move, not because it is the most accurate small size available.
 *
 * ERROR IS NOT MONOTONE IN gridSize: measured, 12 beats 16 (43.87 m vs
 * 44.28 m) and 24 beats 32 (17.70 m vs 17.78 m) on these same control sets,
 * because lattice vertices landing near control points locally cancel
 * error. Nothing here may be read as "denser is always better" — 16 is
 * chosen for its solve/evaluate budget during a drag, not because it beats
 * every smaller size on accuracy.
 */
export const TPS_DRAG_GRID_SIZE = 16;

/**
 * Settled-state tier for a TPS mesh once the pointer stops moving.
 *
 * 32, kept for a third time — now under off-canvas triangle culling and the
 * chunked deferred walk (2026-07-26, Chrome 148 / ANGLE Metal on an Apple
 * M1 Pro — GPU raster confirmed via WEBGL_debug_renderer_info — dpr 2,
 * 1280x800 viewport, the real WarpedRasterLayer on a real Leaflet map,
 * synthetic 7200x5400 ImageBitmap, raster completion forced per frame via
 * the tiny-canvas readback described below; two repeats, spread under 3%).
 *
 * Culling collapsed the walks the drape overhangs: the deep-zoom
 * viewport-capped walk went from 126 to 13–54 ms at 32 and from 466 to
 * 27–36 ms at 64, and a padding-exhausting pan now blits its still-valid
 * backing and blocks for 0–3 ms. Neither path distinguishes the two grids
 * any more. What is left is the peak zoom — the sheet just filling the
 * padded viewport, which is the zoom a sheet is worked at — where nothing
 * culls:
 *
 *     at the peak zoom                    gridSize 32   gridSize 64
 *     deferred refinement, frames              87           327
 *     ...as wall clock at 60 Hz              ~1.5 s        ~5.4 s
 *     worst single refinement frame, ms        33            94
 *     median refinement frame, ms              13            12
 *     edit settle (synchronous), ms            57           208
 *     first display of the sheet, ms          925          3606
 *
 * Chunking removed the freeze but not the work: the walk is now background
 * refinement over a geometrically correct (merely resampled) composite, so
 * the cost reads as TIME-TO-SHARP, not as a stall. At 64 that is ~5.4 s
 * after every zoom settle at the working zoom, against ~1.5 s at 32 — and
 * the edit settle stays synchronous by design (it is the drag tier's
 * uncached hot path), so 64 also blocks ~208 ms on every control-point
 * release against 32's 57 ms. That buys 6.0/1.1/2.0 m of mesh error (max,
 * three real Church control sets) against 32's 17.70–17.78 m, which per the
 * non-monotonicity note above is not even a local optimum — 24 measured
 * slightly better. Not worth a 3.5x longer sharpen and a 3.6x longer edit
 * stall.
 *
 * What would actually change this answer is not a bigger number here: the
 * per-triangle cost is dominated by SOURCE SIZE, not triangle count. Same
 * 2560x2560 destination, same 2048 triangles, varying only the source:
 * 3072x2304 (7.1 Mpx) costs 13 ms, 3600x2700 (9.7 Mpx) costs 587 ms, and
 * 7200x5400 costs 611 ms — a ~45x cliff between roughly 7 and 10 Mpx,
 * where the source stops fitting the GPU image cache and every clipped
 * drawImage re-uploads it. Real scanned sheets are always past that cliff.
 * Drawing the mesh from a resolution-appropriate downscale of the source
 * (an LOD pyramid) would cut every number in the table by more than an
 * order of magnitude and make 64 affordable; that is the next lever, and
 * it is a separate piece of work.
 *
 * Pre-cache history (same rig, `drawWarpedImage` copied into a bench page,
 * 2560x1600 viewport-sized destination, paid on EVERY pan and zoom):
 * 8/16/32/64/128 -> 15.7/54/210/829/3423 ms, LINEAR in triangles — the
 * node-canvas "sublinear" claim did not survive a real browser, and
 * `createImageBitmap(canvas)` resolves before raster completes (2.3 ms vs
 * the real 54 ms), which is why completion is forced by drawing an 8x8
 * snapshot onto a second tiny canvas and reading THAT back. A JS clock
 * inside the walk is blind for the same reason — a chunk reporting 8 ms of
 * elapsed time was measured costing 750 ms of completed GPU work — which
 * is why chunk size is adapted from frame deltas rather than timed
 * in-loop.
 */
export const TPS_GRID_SIZE = 32;

/**
 * Lattice of geographic positions over the raster for a TPS warp, in the
 * same row/col order as `buildGcpLatLngMesh` (row = pixel Y, col = pixel X)
 * so `WarpedRasterLayer` can consume either mesh without caring which
 * solver produced it.
 *
 * Unlike the affine mesh, a TPS surface is not affine anywhere — the
 * bending term means straight lines in pixel space do not stay straight
 * after warping — so a single cell cannot represent it exactly and a real
 * lattice (`TPS_DRAG_GRID_SIZE` while dragging, `TPS_GRID_SIZE` once
 * settled) is required rather than merely offered as an option.
 */
export function buildTpsLatLngMesh(
  params: TpsParams,
  pixelSize: PixelSize,
  gridSize: number = TPS_GRID_SIZE,
  sourceRect?: PixelRect,
): LatLngPoint[][] {
  const rect = resolveSourceRect(pixelSize, sourceRect);
  const mesh: LatLngPoint[][] = [];
  for (let row = 0; row <= gridSize; row += 1) {
    const line: LatLngPoint[] = [];
    const y = rect.y + (rect.height * row) / gridSize;
    for (let col = 0; col <= gridSize; col += 1) {
      const x = rect.x + (rect.width * col) / gridSize;
      line.push(fromMercator(applyTps(params, x, y)));
    }
    mesh.push(line);
  }
  return mesh;
}

File: NSMarksCore/Sources/GeoCore/Georeference/GcpMesh.swift
import Foundation

/// The raster's size in its own pixels.
public struct PixelSize: Hashable, Sendable {
    public var width: Double
    public var height: Double

    public init(width: Double, height: Double) {
        self.width = width
        self.height = height
    }
}

/// A rectangle of the raster, in the same pixels — the part of a scan the user
/// kept, when they cropped the margins off a sheet.
public struct PixelRect: Hashable, Sendable {
    public var x: Double
    public var y: Double
    public var width: Double
    public var height: Double

    public init(x: Double, y: Double, width: Double, height: Double) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

/// Lattices of ground positions over a raster, one vertex per grid node.
///
/// Ported from `web/src/userMaps/transform/gcpMesh.ts`, including the two grid
/// sizes and the reasons they are what they are. The measurements quoted in
/// those comments were taken in a browser and are carried across as the record
/// of why the numbers were chosen; they are not claims about this renderer,
/// which draws the mesh through a different path and will need its own.
public enum GcpMesh {
    /// Why a lattice cannot be built.
    public enum MeshRefusal: Error, Equatable, Sendable {
        /// Not a rectangle: a non-finite edge, or no area.
        case degenerate
        /// Asks for pixels the raster does not have.
        case outsideRaster
        /// A grid size below one. Silently rounding it up to a single cell
        /// would throw away a spline's whole bending term — a hundred metres
        /// on a county-scale sheet — and report success.
        case notALattice
    }

    /// How far outside the raster an edge may sit and still be read as sitting
    /// on it. A crop expressed as a fraction of the width lands a hair past the
    /// last pixel often enough that refusing it would refuse ordinary work.
    static let edgeTolerancePixels = 1e-7

    /// One cell, two triangles — and that is exact, not an approximation.
    ///
    /// The screen's space is Web Mercator scaled and translated, so a
    /// pixel-to-Mercator affine composes with it into a map that is still
    /// affine all the way to the canvas. There is no curvature for a denser
    /// lattice to absorb. (An embedded-GeoTIFF path needs more, because it goes
    /// pixel to UTM to WGS 84 to Mercator, and UTM to Mercator genuinely
    /// curves.)
    ///
    /// It is also the performance answer: the georeferencer re-solves on every
    /// touch move during a drag, and each cell costs two clipped draws over the
    /// whole preview.
    public static let affineGridSize = 1

    /// Coarse tier for a spline mesh while a control point is being dragged.
    ///
    /// On the web, at 16 the mesh error measured 44.3/10.9/15.9 ground metres
    /// (max, over three real Church control sets) and re-solving stayed inside
    /// a 16 ms frame up to 300 control points. That budget is what chose 16.
    ///
    /// Error is not monotone in grid size: measured, 12 beat 16 (43.87 m vs
    /// 44.28 m) and 24 beat 32 (17.70 m vs 17.78 m) on those same sets, because
    /// lattice vertices landing near control points locally cancel error.
    /// Nothing here may be read as "denser is always better".
    public static let splineDragGridSize = 16

    /// Settled tier, once the pointer stops moving.
    ///
    /// On the web, 64 bought 6.0/1.1/2.0 m of mesh error against 32's
    /// 17.70–17.78 m and cost about 3.5x the time-to-sharp after every zoom
    /// settle and about 3.6x the block on every control-point release. Per the
    /// non-monotonicity above, 64 is not even a local optimum — 24 measured
    /// slightly better.
    ///
    /// What would actually move this number is not the number: per-triangle
    /// cost is dominated by source size, not triangle count. Same destination
    /// and same triangle count, varying only the source, the web measured
    /// 7.1 Mpx at 13 ms against 9.7 Mpx at 587 ms — a roughly 45x cliff where
    /// the source stops fitting the GPU image cache and every clipped draw
    /// re-uploads it. Real scanned sheets are always past that cliff, so
    /// drawing from a resolution-appropriate downscale is the lever, and it is
    /// separate work.
    public static let splineGridSize = 32

    /// The rectangle to evaluate over: the one given, checked against the
    /// raster, or the whole raster when none was given.
    public static func resolve(
        pixelSize: PixelSize, sourceRect: PixelRect?
    ) throws(MeshRefusal) -> PixelRect {
        let rect = sourceRect ?? PixelRect(
            x: 0, y: 0, width: pixelSize.width, height: pixelSize.height
        )
        let edges = [rect.x, rect.y, rect.width, rect.height,
                     pixelSize.width, pixelSize.height]
        guard edges.allSatisfy(\.isFinite), rect.width > 0, rect.height > 0 else {
            throw .degenerate
        }
        guard rect.x >= -Self.edgeTolerancePixels,
              rect.y >= -Self.edgeTolerancePixels,
              rect.x + rect.width <= pixelSize.width + Self.edgeTolerancePixels,
              rect.y + rect.height <= pixelSize.height + Self.edgeTolerancePixels
        else {
            throw .outsideRaster
        }
        // Clamped rather than trusted: the tolerance above admits an edge a
        // hair outside the raster, and sampling there would read pixels that
        // do not exist.
        let x = min(max(0, rect.x), pixelSize.width)
        let y = min(max(0, rect.y), pixelSize.height)
        let width = min(rect.width, pixelSize.width - x)
        let height = min(rect.height, pixelSize.height - y)
        // A crop narrower than the tolerance, sitting a hair past the right or
        // bottom edge, passes the check above and clamps to a negative extent —
        // a lattice that runs backwards out of the raster. There is no
        // rectangle there to evaluate, so it is refused rather than drawn.
        guard width > 0, height > 0 else { throw .outsideRaster }
        return PixelRect(x: x, y: y, width: width, height: height)
    }

    /// Lattice of ground positions for an affine warp.
    ///
    /// Row is pixel Y and column is pixel X, so a renderer can consume either
    /// this or the spline mesh without knowing which solver produced it.
    public static func latLngMesh(
        _ transform: AffineTransform,
        pixelSize: PixelSize,
        gridSize: Int = affineGridSize,
        sourceRect: PixelRect? = nil
    ) throws(MeshRefusal) -> [[GeoPoint]] {
        try mesh(pixelSize: pixelSize, gridSize: gridSize, sourceRect: sourceRect) {
            transform.apply(x: $0, y: $1)
        }
    }

    /// Lattice of ground positions for a spline warp.
    ///
    /// Unlike the affine mesh, a spline surface is not affine anywhere — the
    /// bending term means straight lines in pixel space do not stay straight —
    /// so one cell cannot represent it and a real lattice is required rather
    /// than merely offered.
    public static func latLngMesh(
        _ spline: ThinPlateSpline,
        pixelSize: PixelSize,
        gridSize: Int = splineGridSize,
        sourceRect: PixelRect? = nil
    ) throws(MeshRefusal) -> [[GeoPoint]] {
        try mesh(pixelSize: pixelSize, gridSize: gridSize, sourceRect: sourceRect) {
            spline.apply(x: $0, y: $1)
        }
    }

    private static func mesh(
        pixelSize: PixelSize,
        gridSize: Int,
        sourceRect: PixelRect?,
        project: (Double, Double) -> MercatorPoint
    ) throws(MeshRefusal) -> [[GeoPoint]] {
        // A lattice needs at least the one cell the affine tier uses; zero
        // would divide by zero and a negative would produce nothing at all.
        guard gridSize >= 1 else { throw .notALattice }
        let rect = try resolve(pixelSize: pixelSize, sourceRect: sourceRect)
        let steps = gridSize
        return (0...steps).map { row in
            let y = rect.y + rect.height * Double(row) / Double(steps)
            return (0...steps).map { col in
                let x = rect.x + rect.width * Double(col) / Double(steps)
                return WebMercator.unproject(project(x, y))
            }
        }
    }
}
